man/adoc/bpftrace.adoc

```diff
@@ -0,0 +1,1500 @@
+
+Default: 64
+
+Number of pages to allocate per CPU for perf ring buffer.
+The value must be a power of 2.
+
+If you're getting a lot of dropped events bpftrace may not be processing events in the ring buffer fast enough.
+It may be useful to bump the value higher so more events can be queued up.
+The tradeoff is that bpftrace will use more memory.
+
+== Supported architectures
+
+x86, arm64 and s390x
+
+//TODO: is this true?
+
+== bpftrace Language
+
+=== Overview
+
+The `bpftrace` (`bt`) language is inspired by the D language used by `dtrace` and uses the same program structure.
+Each script consists of an preamble and one or more action blocks.
+
+----
+preamble
+
+actionblock1
+actionblock2
+----
+
+Preprocessor and type definitions take place in the preamble:
+
+----
+#include <linux/socket.h>
+#define RED "\033[31m"
+
+struct S {
+  int x;
+}
+----
+
+
+Each action block consists of three parts:
+
+----
+probe[,probe]
+/predicate/ {
+  action
+}
+----
+
+Probes::
+  A probe specifies the event and event type to attach too.
+
+Predicate::
+  The predicate is optional condition that must be met for the action to be executed.
+
+Action::
+  Actions are the programs that run when an event fires (and the predicate is met).
+An action is a semicolon (`;`) separated list of statements and always enclosed by brackets `{}`
+
+A basic script that traces the `open(2)` and `openat(2)` system calls can be written as follows:
+
+----
+BEGIN
+{
+	printf("Tracing open syscalls... Hit Ctrl-C to end.\n");
+}
+
+tracepoint:syscalls:sys_enter_open,
+tracepoint:syscalls:sys_enter_openat
+{
+	printf("%-6d %-16s %s\n", pid, comm, str(args->filename));
+}
+----
+
+This script has two action blocks and a total of 3 probes.
+The first action block uses the special `BEGIN` probe, which fires once during `bpftrace` startup.
+This probe is used to print a header that indicates that the tracing has started.
+
+The second action block uses two probes, one for `open` and one for `openat`, and defines an action that prints the file being `open`ed as well as the `pid` and `comm` of the process that execute the syscall.
+
+=== Identifiers
+
+Identifiers must match the following regular expression: `[_a-zA-Z][_a-zA-Z0-9]*`
+
+=== Comments
+
+Both single line and multi line comments are supported.
+
+----
+// A single line comment
+i:s:1 { // can also be used to comment inline
+/*
+ a multi line comment
+
+*/
+  print(/* inline comment block */ 1);
+}
+----
+
+=== Data Types
+
+The following fundamental integer types are provided by the language.
+
+[cols="~,~"]
+|===
+|*Type*
+|*Description*
+
+|uint8
+|Unsigned 8 bit integer
+
+|int8
+|Signed 8 bit integer
+
+|uint16
+|Unsigned 16 bit integer
+
+|int16
+|Signed 16 bit integer
+
+|uint32
+|Unsigned 32 bit integer
+
+|int32
+|Signed 32 bit integer
+
+|uint64
+|Unsigned 64 bit integer
+
+|int64
+|Signed 64 bit integer
+|===
+
+==== Floating-point
+
+Floating-point numbers are not supported by BPF and therefor not by bpftrace.
```

Comment:
```suggestion
Floating-point numbers are not supported by BPF and therefore not by bpftrace.
```
Actually I think they're adding support: https://lore.kernel.org/bpf/20210226202256.116518-1-iii@linux.ibm.com/ . Unsure what it actually enables, though

Reply:
this seems to be btf only